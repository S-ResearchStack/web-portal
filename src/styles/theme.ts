import { css, DefaultTheme } from 'styled-components';
import { px } from './utils';

const colors = {
  // gray gradations
  black: '#000000',
  black60: 'rgba(0, 0, 0, 0.6)',
  black40: 'rgba(0, 0, 0, 0.4)',
  black08: 'rgba(0, 0, 0, 0.08)',
  black15: 'rgba(0, 0, 0, 0.15)',

  // blue gradations
  blue: "#b0c4f1",
  blue40: 'rgba(176, 196, 241, 0.4)',
  blue15: 'rgba(176, 196, 241, 0.15)',
  blue8: 'rgba(176, 196, 241, 0.08)',

  // light mode
  surface: '#FFFFFF',
  background: '#F7F8FA',
  onPrimary: '#FFFFFF',
  backgroundOnPrimary: '#FFFFFF',
  backgroundSurface: '#FFFFFF',
  backgroundOnSurface: '#2D2D2D',
  primary: '#4475E3',
  primaryShadow: '#4475E361',
  primaryLight: '#ECF1FC',
  primaryHovered: '#2F65E0',
  primary05: '#F6F8FE',
  primary10: '#ECF1FC',
  primary20: '#E3EAFB',
  primary30: '#C7D6F7',
  primaryDisabled: '#B3C6F1',
  primaryLightPressed: '#E3EAFB',
  primaryLightFocused: '#C7D6F7',
  textPrimary: '#474747',
  primaryWhite: '#FFFFFF',
  disabled: '#DBDBDB',
  disabled20: 'rgba(219, 219, 219, 0.2)',
  onDisabled: '#A1A1A1',

  textPrimaryBlue: '#4475E3',
  textPrimaryBlue10: 'rgba(68, 117, 227, 0.2)',
  textPrimaryDark: '#474747',
  textSecondaryGray: '#808080',
  textDisabled: '#838383',
  onSurface: '#2D2D2D',
  secondarySkyblue: '#00B0D7', // replace
  backgoundLight: '#F0F0F0',
  primaryBlueHovered: '#2F65E0',
  primaryBluePressed: '#2057D5',

  statusSuccess: '#2DC008',
  statusSuccess10: '#DCF2EA',
  statusSuccessText: '#00825D',

  statusError: '#FF3333',
  statusError10: '#FFEBEB',
  statusErrorText: '#D14343',

  statusWarning: '#F5BF00',
  statusWarning10: '#FFF7D0',
  statusWarningText: '#A36400',

  secondaryViolet: '#944ED7',
  secondaryVioletHovered: '#6A3799',
  secondaryViolet10: '#F4EDFB',

  secondaryTangerine: '#F39C18',
  secondaryTangerineHovered: '#CC8314',
  secondaryTangerine10: '#FEF5E8',

  secondarySkyBlue: '#00B0D7',
  secondarySkyBlueHovered: '#0092B3',
  secondarySkyBlue10: '#E5F7FB',

  secondaryGreen: '#1AD598',
  secondaryGreenHovered: '#16B37F',
  secondaryGreen10: '#E8FBF5',

  secondaryRed: '#FA5F4F',
  secondaryRedHovered: '#CC4E40',
  secondaryRed10: '#FFEFED',

  transparent: 'rgba(0, 0, 0, 0)'
};

export type SpecColorType = keyof typeof colors;

const weightToCssValue = {
  regular: 400,
  upperregular: 450,
  medium: 500,
  semibold: 600,
  bold: 700,
};

const fontSpec = (
  size: number,
  lineHeight: number | string,
  weight: keyof typeof weightToCssValue,
  fontFamily: string
) => css`
  font-family: ${fontFamily}, sans-serif;
  font-size: ${px(size)};
  line-height: ${typeof lineHeight === 'string' ? lineHeight : px(lineHeight)};
  font-weight: ${weightToCssValue[weight]};
  font-style: normal;
`;

const typography = {
  // Inter typeface
  labelRegular: fontSpec(10, 15, 'regular', 'Inter'),
  labelSemibold: fontSpec(10, 15, 'semibold', 'Inter'),
  smallLabelRegular: fontSpec(8, 10, 'regular', 'Inter'),

  bodyXSmallRegular: fontSpec(12, 18, 'regular', 'Inter'),
  bodyXSmallSemibold: fontSpec(12, 18, 'semibold', 'Inter'),
  bodySmallRegular: fontSpec(14, 21, 'regular', 'Inter'),
  bodySmallSemibold: fontSpec(14, 21, 'semibold', 'Inter'),
  bodyMediumRegular: fontSpec(16, 24, 'regular', 'Inter'),
  bodyMediumSemibold: fontSpec(16, 24, 'semibold', 'Inter'),
  bodyLargeRegular: fontSpec(18, 26, 'regular', 'Inter'),
  bodyLargeSemibold: fontSpec(18, 26, 'semibold', 'Inter'),

  headingXSmall: fontSpec(16, 21, 'semibold', 'Inter'),
  headingSmall: fontSpec(18, 23, 'semibold', 'Inter'),
  headingXMedium: fontSpec(18, 27, 'semibold', 'Inter'),
  headingXMediumRegular: fontSpec(18, 27, 'regular', 'Inter'),
  headingMedium: fontSpec(24, 32.2, 'semibold', 'Inter'),
  headingLargeSemibold: fontSpec(32, 42, 'semibold', 'Inter'),
  heading2LargeSemibold: fontSpec(40, 52, 'semibold', 'Inter'),
  headingXLargeSemibold: fontSpec(48, 58, 'semibold', 'Inter'),
  headingXXLargeSemibold: fontSpec(60, 72, 'semibold', 'Inter'),

  portalDisplayLarge: fontSpec(72, 86, 'semibold', 'Inter'),

  sdkBodySmallRegular: fontSpec(14, '130%', 'regular', 'Inter'),
  sdkBodyLargeRegular: fontSpec(18, 24, 'regular', 'Inter'),
  sdkBodyLargeSemibold: fontSpec(18, 24, 'semibold', 'Inter'),
  sdkHeadingMedium: fontSpec(20, 26, 'semibold', 'Inter'),
  sdkBodyXXLRegular: fontSpec(32, 42, 'regular', 'Inter'),
  sdkBodyXXLSemibold: fontSpec(32, '130%', 'semibold', 'Inter'),
  sdkBodyMediumRegular: fontSpec(16, '130%', 'regular', 'Inter'),
  sdkBodyMediumSemibold: fontSpec(16, '130%', 'semibold', 'Inter'),
  sdkDisplayMedium: fontSpec(20, 26, 'semibold', 'Inter'),
  sdkHeadingLarge: fontSpec(40, '130%', 'semibold', 'Inter'),

  // Fira Code typeface
  query14: fontSpec(14, '130%', 'upperregular', 'Fira Code'),
  query12: fontSpec(12, '150%', 'upperregular', 'Fira Code'),
};

const boxShadow = {
  card: `0 0 ${px(2)} ${px(0)} rgba(0,0,0,0.15)`,
  avatar: `0 ${px(8)} ${px(10)} ${px(2)} 0 rgba(0,0,0,0.12)`,
  studyAvatar: `0 ${px(15)} ${px(20)} rgba(180,180,180,0.3)`,
  previewScreen: `0 ${px(2)} ${px(4)} rgba(71,71,71,0.25)`,
  skipLogicHeader: `box-shadow: ${px(3)} ${px(4)} ${px(15)} rgba(0, 0, 0, 0.05)`,
};

const animation = {
  defaultTiming: 'cubic-bezier(0.45, 0.05, 0, 1)',
  rippleTiming: 'cubic-bezier(0.4, 0, 0.2, 1)',
};

declare module 'styled-components' {
  export interface DefaultTheme {
    colors: typeof colors;
    typography: typeof typography;
    boxShadow: typeof boxShadow;
    animation: typeof animation;
  }
}

const theme: DefaultTheme = {
  colors,
  typography,
  boxShadow,
  animation,
};

export default theme;
