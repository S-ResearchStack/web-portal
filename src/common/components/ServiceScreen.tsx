import React, { FC, ReactElement } from 'react';
import useMeasure from 'react-use/lib/useMeasure';
import styled from 'styled-components';

import Button from 'src/common/components/Button';
import Spinner from 'src/common/components/Spinner';

import ErrorServiceIcon from 'src/assets/service/error.svg';
import EmptyServiceIcon from 'src/assets/service/empty_data.svg';
import { px, typography } from 'src/styles';

const COMPACT_MIN_WIDTH = 420;

const COMPACT_MIN_HEIGHT = 400;

enum ServiceScreenType {
  error = 'error',
  empty = 'empty',
  loading = 'loading',
}

interface ServiceScreenPreparedProps {
  icon?: React.FunctionComponent<React.SVGProps<SVGSVGElement>>;
  title?: React.ReactNode;
}

interface ServiceScreenProps
  extends React.PropsWithChildren<
    Omit<ServiceScreenPreparedProps, 'icon'> & Omit<React.HTMLAttributes<HTMLDivElement>, 'title'>
  > {
  type: keyof typeof ServiceScreenType;
  title?: React.ReactNode;
  onReload?: () => void;
}

const CONTAINER_PADDING = 32;

const Container = styled.div`
  box-sizing: border-box;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
  min-width: 100%;
  min-height: 100%;
  flex: 1;
`;

const Image = styled.div<{ $empty?: boolean }>`
  margin-bottom: ${({ $empty }) => ($empty ? 0 : px(18))};
`;

const Title = styled.div<{ $empty?: boolean }>`
  ${typography.bodySmallRegular};
  color: ${({ $empty, theme }) => ($empty ? theme.colors.textPrimary : theme.colors.onSurface)};
  margin-bottom: ${px(23)};
`;

const ButtonContainer = styled.div`
  margin-top: ${px(69)};
  width: ${px(248)};
`;

const ServiceScreen: FC<ServiceScreenProps> = ({
  type,
  title,
  children,
  onReload,
  ...props
}): ReactElement => {
  const [containerRef, { width, height }] = useMeasure<HTMLDivElement>();

  const isCompact =
    width + CONTAINER_PADDING * 2 < COMPACT_MIN_WIDTH ||
    height + CONTAINER_PADDING * 2 < COMPACT_MIN_HEIGHT;

  const renderScreen = () => {
    if (type === 'error') {
      return (
        <>
          {!isCompact && (
            <Image>
              <ErrorServiceIcon />
            </Image>
          )}
          <Title>{title || 'Server Error'}</Title>
          {onReload && (
            <ButtonContainer>
              <Button fill="solid" onClick={onReload} data-testid="reload-button">
                Reload
              </Button>
            </ButtonContainer>
          )}
        </>
      );
    }
    if (type === 'empty' || type === 'loading') {
      let icon;

      if (type === 'loading') {
        icon = <Spinner size="l" data-testid="spinner" />;
      } else if (type === 'empty') {
        icon = <EmptyServiceIcon data-testid="empty-icon" />;
      }

      return (
        <>
          <Image $empty>{icon}</Image>
          <Title $empty>{title || 'No Data'}</Title>
        </>
      );
    }

    return null;
  };

  return (
    <Container ref={containerRef} {...props}>
      {renderScreen()}
    </Container>
  );
};

export default ServiceScreen;
