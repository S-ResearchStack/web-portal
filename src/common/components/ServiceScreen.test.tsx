import React from 'react';
import '@testing-library/jest-dom';
import 'jest-styled-components';
import { fireEvent, getByText, render } from '@testing-library/react';
import theme from 'src/styles/theme';
import { ThemeProvider } from 'styled-components/';
import ServiceScreen from './ServiceScreen';

describe('ServiceScreen', () => {
  it('[NEGATIVE] test error service screen render', async () => {
    const onReload = jest.fn();

    const { baseElement, getByTestId, queryByTestId } = render(
      <ThemeProvider theme={theme}>
        <ServiceScreen type="error" onReload={onReload} data-testid="error-screen" />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();

    const errorScreen = getByTestId('error-screen');
    const reloadButton = queryByTestId('reload-button') as Element;

    expect(errorScreen).toBeInTheDocument();
    expect(reloadButton).toBeInTheDocument();

    fireEvent.click(reloadButton);

    expect(onReload).toHaveBeenCalledTimes(1);
  });

  it('[NEGATIVE] test empty service screen render', () => {
    const { baseElement, getByTestId, queryByTestId } = render(
      <ThemeProvider theme={theme}>
        <ServiceScreen type="empty" data-testid="empty-screen" />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();

    const emptyScreen = getByTestId('empty-screen');
    const emptyIcon = queryByTestId('empty-icon') as Element;

    expect(emptyScreen).toBeInTheDocument();
    expect(emptyIcon).toBeInTheDocument();
    expect(getByText(baseElement, 'No Data')).toBeInTheDocument();
  });

  it('test loading service screen render', () => {
    const { baseElement, getByTestId, queryByTestId } = render(
      <ThemeProvider theme={theme}>
        <ServiceScreen type="loading" data-testid="loading-screen" />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();

    const errorScreen = getByTestId('loading-screen');
    const spinner = queryByTestId('spinner') as Element;

    expect(errorScreen).toBeInTheDocument();
    expect(spinner).toBeInTheDocument();
  });

  it('[NEGATIVE] should render with wrong `title` property', () => {
    const { baseElement, getByTestId } = render(
      <ThemeProvider theme={theme}>
        <ServiceScreen type="error" title={false as unknown as string} data-testid="error-screen" />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();
    expect(getByTestId('error-screen')).toBeInTheDocument();
    expect(getByText(baseElement, 'Server Error')).not.toBeNull();
  });

  it('[NEGATIVE] should render with wrong `type` property', () => {
    const { baseElement, getByTestId } = render(
      <ThemeProvider theme={theme}>
        <ServiceScreen type={'unknown' as 'error'} data-testid="error-screen" />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();
    expect(getByTestId('error-screen')).toBeInTheDocument();
  });
});
