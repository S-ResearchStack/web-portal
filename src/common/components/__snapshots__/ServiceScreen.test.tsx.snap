// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceScreen [NEGATIVE] should render with wrong \`title\` property 1`] = `
.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  min-width: 100%;
  min-height: 100%;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 400;
  font-style: normal;
  color: #2D2D2D;
  margin-bottom: 1.4375rem;
}

<body>
  <div>
    <div
      class="c0"
      data-testid="error-screen"
    >
      <div
        class="c1"
      >
        Server Error
      </div>
    </div>
  </div>
</body>
`;

exports[`ServiceScreen [NEGATIVE] should render with wrong \`type\` property 1`] = `
.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  min-width: 100%;
  min-height: 100%;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<body>
  <div>
    <div
      class="c0"
      data-testid="error-screen"
    />
  </div>
</body>
`;

exports[`ServiceScreen [NEGATIVE] test empty service screen render 1`] = `
.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  min-width: 100%;
  min-height: 100%;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 {
  margin-bottom: 0;
}

.c2 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 400;
  font-style: normal;
  color: #474747;
  margin-bottom: 1.4375rem;
}

<body>
  <div>
    <div
      class="c0"
      data-testid="empty-screen"
    >
      <div
        class="c1"
      >
        <svg
          data-testid="empty-icon"
        />
      </div>
      <div
        class="c2"
      >
        No Data
      </div>
    </div>
  </div>
</body>
`;

exports[`ServiceScreen [NEGATIVE] test error service screen render 1`] = `
.c5 {
  position: absolute;
  will-change: opacity,transform;
  border-radius: 50%;
  -webkit-animation: bykygw 600ms cubic-bezier(0.4,0,0.2,1) forwards;
  animation: bykygw 600ms cubic-bezier(0.4,0,0.2,1) forwards;
  z-index: 0;
  pointer-events: none;
}

.c3 {
  position: relative;
  z-index: 0;
  display: -webkit-box !important;
  display: -webkit-flex !important;
  display: -ms-flexbox !important;
  display: flex !important;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-radius: 0.25rem;
  width: 100%;
  white-space: nowrap;
  padding: 0;
  -webkit-transition-duration: 400ms;
  transition-duration: 400ms;
  -webkit-transition-property: background-color;
  transition-property: background-color;
  height: 3rem;
  border: none;
  background-color: #4475E3;
}

.c3:hover {
  cursor: pointer;
}

.c3 > div {
  font-family: Inter,sans-serif;
  font-size: 1rem;
  line-height: 1.5rem;
  font-weight: 600;
  font-style: normal;
}

.c3 > div svg {
  margin-right: 0.25rem;
}

.c3 > div {
  color: #FFFFFF;
}

.c3 > div svg {
  fill: #FFFFFF;
}

.c3:hover:enabled {
  background-color: #2F65E0;
}

.c3:disabled {
  background-color: #B3C6F1;
}

.c3:focus-visible:enabled {
  border: 0.125rem solid #FFFFFF;
  background-color: #4475E3;
  outline: 0.125rem solid #C7D6F7;
}

.c4 {
  position: relative;
  z-index: 2;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transform: translate3d(0,0,0.0625rem);
  -ms-transform: translate3d(0,0,0.0625rem);
  transform: translate3d(0,0,0.0625rem);
}

.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  min-width: 100%;
  min-height: 100%;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 400;
  font-style: normal;
  color: #2D2D2D;
  margin-bottom: 1.4375rem;
}

.c2 {
  margin-top: 4.3125rem;
  width: 15.5rem;
}

<body>
  <div>
    <div
      class="c0"
      data-testid="error-screen"
    >
      <div
        class="c1"
      >
        Server Error
      </div>
      <div
        class="c2"
      >
        <button
          class="c3"
          data-testid="reload-button"
          fill="solid"
          style="overflow: hidden; position: relative;"
        >
          <div
            class="c4"
            data-testid="content"
          >
            Reload
          </div>
          <div
            class="c5"
            color="primaryPressed"
            opacity="1,1"
            scale="0"
            style="display: none; width: 0rem; height: 0rem; top: 0rem; left: 0rem;"
          />
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`ServiceScreen test loading service screen render 1`] = `
.c2 {
  fill: #4475E3;
  -webkit-animation: 1.1s infinite cilQsd cubic-bezier(0.45,0.05,0,1);
  animation: 1.1s infinite cilQsd cubic-bezier(0.45,0.05,0,1);
}

.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  min-width: 100%;
  min-height: 100%;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 {
  margin-bottom: 0;
}

.c3 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 400;
  font-style: normal;
  color: #474747;
  margin-bottom: 1.4375rem;
}

<body>
  <div>
    <div
      class="c0"
      data-testid="loading-screen"
    >
      <div
        class="c1"
      >
        <svg
          class="c2"
          data-testid="spinner"
        />
      </div>
      <div
        class="c3"
      >
        No Data
      </div>
    </div>
  </div>
</body>
`;
